docs/adr/0008-sync-payload-evolution.md

```diff
@@ -0,0 +1,144 @@
+# Handling the evolution of Sync payloads
+
+* Status: Accepted
+* Deciders: sync team, credentials management team
+* Date: 2023-03-15
+
+Technical Story:
+* https://github.com/mozilla/application-services/pull/5434
+* https://docs.google.com/document/d/1ToLOERA5HKzEzRVZNv6Ohv_2wZaujW69pVb1Kef2jNY
+
+## Context and Problem Statement
+
+Sync exists on all platforms (Desktop, Android, iOS), all channels (Nightly, Beta, Release, ESR) and is heavily used across all Firefox features.
+Whenever there are feature changes or requests that potentially involve schema changes, there are not a lot of good options to ensure sync doesn’t break for any specific client.
+Since sync data is synced from all channels, we need to make sure each client can handle the new data and that all channels can support the new schema.
+Issues like [credit card failing on android and desktop release channels due to schema change on desktop Nightly](https://bugzilla.mozilla.org/show_bug.cgi?id=1812235)
+is an example of such cases we can run into.
+This document describes our decision on how we will support payload evolution over time.
+
+Note that even though this document existing in the application-services repoository, it should
+be considered to apply too all sync implementation, whether in this repository, in mozilla-central,
+or anywhere else.
+
+## Definitions
+
+* A "new" Firefox installation is a version of Firefox which has a change to a Sync payload which
+  is not yet understood by "recent" versions. The most common example would be a Nightly version
+  of Firefox with a new feature not yet on the release channel.
+
+* A "recent" Firefox installation is a version older than a "new" version, which does not understand
+  or have support for new features in "new" versions, but which we still want to support without
+  breakage and without the user perceiving data-loss. This is typically accepted to mean the
+  current ESR version or later, but taking into account the slow update when new ESRs are released.
+
+* An "old" version is any version before what we consider "recent".
+
+
+## Decision Drivers
+
+* It must be possible to change what data is carried by Sync to meet future product requirements.
+* Both desktop and mobile platforms must be considered.
+* We must not break "recent" Firefox installations, nor cause what the user might perceive as data-loss,
+  when a "new" Firefox installation syncs, and vice-versa.
+* Some degree of breakage for "old" Firefox installations when "new" or "recent" firefoxes sync
+  might be considered acceptable if absolutely necessary.
+* However, breakage of "new" or "recent" Firefoxes should "old" versions sync is *not* acceptable.
+* Because such evolution should be rare, we do not want to set an up-front policy about locking out
+  "old" versions just because they might have a problem in the future. That is, we want to avoid
+  a policy that dictates versions more than (say) 2 years old will break when syncing "just in case"
```

Comment:
I actually don't like "decision drivers" because I don't quite understand what it means :) If I renamed this to simply "requirements" (which I do :) then I think this para still belongs there, right? Or are you suggesting that it's misplaced because it's really just stating a non-requirement?

Reply:
The way I see it "Some degree of breakage for "old" Firefox installations when "new" or "recent" firefoxes sync might be considered acceptable if absolutely necessary." is the requirement, which is why we don't want to have a blanket policy of locking out older versions.  I would either remove the bullet or discuss locking out older versions as in the options section.  But I'm definitely open to keeping it here if you like it.